### PR TITLE
Add download functionnality

### PR DESCRIPTION
**Pull request by vincent-boulin**
Hello,

I'm user of your grunt plugin grunt-ftp, and for some projects we need to put and get some files in an FTP Server.
Like i don't find some grunt plugin which do the both, i add this functionnality into your plugin.

Let me know if what i've done is great or if it's need to change some other things.

I don't have think of comptability with older version, so maybe we will need to rename ftpPut task to its older name.

PS : Sorry my english, i'm french.

Regards,

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "grunt-ftp",
-  "version": "1.0.2",
+  "version": "1.1.0",
```
Comment (sindresorhus) [inline package.json]: don't bump

File: tasks/ftp.js [hunk 2]
```diff
@@ -56,4 +56,58 @@ module.exports = function (grunt) {
...
+		if (options.host === undefined) {
+			throw new Error('`host` required');
+		}
+
+
+		eachAsync(this.files, function (el, i, next) {
+			// have to create a new connection for each file otherwise they conflict
+			var ftp = new JSFtp(options);
+
+			if(!grunt.file.exists(el.dest)) {
```
Comment (sindresorhus) [inline tasks/ftp.js]: No need to `if` check it, just use `grunt.file.mkdir` directly.

File: readme.md [hunk 3]
```diff
@@ -35,6 +35,28 @@ grunt.initConfig({
...
+		options: {
+			host: 'website.com',
+			user: 'johndoe',
+			pass: '1234'
+		},
+		upload: {
+			files: {
+				'public_html': 'src/file.txt'
+			}
+		}
```
Comment (sindresorhus) [inline readme.md]: this needs to be updated

File: readme.md [hunk 4]
```diff
@@ -35,6 +35,28 @@ grunt.initConfig({
...
+	}
+});
+
+grunt.registerTask('default', ['ftp']);
```
Comment (sindresorhus) [inline readme.md]: same

File: readme.md [hunk 5]
```diff
@@ -58,6 +80,9 @@ Default: `'anonymous'`
 Type: `string`  
 Default: `'@anonymous'`
 
+## Notes
+
+For downloading task, if dest path is a folder that doesn't exists, this will be create
```
Comment (sindresorhus) [inline readme.md]: not needed

File: tasks/ftp.js [hunk 6]
```diff
@@ -56,4 +56,58 @@ module.exports = function (grunt) {
...
+				// if dest path doesn't exists, we create the full dirname
+				grunt.file.mkdir(path.dirname(el.dest));
+			}
+
+			var finalLocalPath = el.dest;
+			if(grunt.file.isDir(el.dest)) {
+				// if dest is a directory, have to create a file with source filename
+				var filename = path.basename(el.src[0]);
+				finalLocalPath = path.join(el.dest, '/', filename);
+			}
```
Comment (sindresorhus) [inline tasks/ftp.js]:
```js

**Comment by sk4ry** [comment]
Hello,

I've considered your remarks.
Let me know, if you want me to update the ftpPut task for the backward compatibility.

Thank you.

Regards,

**Comment by sindresorhus** [comment]
> Let me know, if you want me to update the ftpPut task for the backward compatibility.

No need. I'll just alias it back after merging this.

File: tasks/ftp.js [hunk 7]
```diff
@@ -56,4 +56,55 @@ module.exports = function (grunt) {
...
+
+
+		eachAsync(this.files, function (el, i, next) {
+			// have to create a new connection for each file otherwise they conflict
+			var ftp = new JSFtp(options);
+
+			grunt.file.mkdir(path.dirname(el.dest));
+
+			var finalLocalPath = el.dest;
+			if(grunt.file.isDir(el.dest)) {
```
Comment (sindresorhus) [inline tasks/ftp.js]: `if (grunt.file.isDir(el.dest)) {`

File: test.js [hunk 8]
```diff
@@ -5,3 +5,12 @@ var grunt = require('grunt');
 it('should upload files to an FTP-server', function () {
 	assert(grunt.file.exists('ftp/fixture/fixture.txt'));
 });
+
+it('should create a new folder for retrieve files', function() {
```
Comment (sindresorhus) [inline test.js]: `function () {`

File: tasks/ftp.js (hunk 7 above)
Comment (sk4ry) [inline tasks/ftp.js]: This test is used for this config example :
```js
files: {
	'fixtureGet/fixture2.txt': 'ftp/fixture/fixture.txt',
	'<%= dirs.local %>/fixtureGet': 'ftp/fixture/fixture.txt'
}
```

If the dest param is a directory, we create a new file on path : "dest_dir/src_filename".
May be there is a better way to do this ?

File: test.js (hunk 8 above)
Comment (sk4ry) [inline test.js]: Is the test of folder creation useless ?

File: tasks/ftp.js (hunk 7 above)
Comment (sindresorhus) [inline tasks/ftp.js]: @sk4ry I was only commenting on the code style. Notice the difference.

File: test.js (hunk 8 above)
Comment (sindresorhus) [inline test.js]: Code style.

File: tasks/ftp.js (hunk 7 above)
Comment (sk4ry) [inline tasks/ftp.js]: Oh ok, i don't notice it first time.
I'll commit this so.

**Comment by sk4ry** [comment]
Normally it's good now.

Just a question for my knowledge, why do you say me not to bump the version ?

**Comment by sindresorhus** [comment]
> Just a question for my knowledge, why do you say me not to bump the version ?

Because it's not up to you what version I will release it as. I might want to do more changes before a release. I might have a release script which bumps it automagically (I actually do). Just don't bump other peoples versions ;)

**Comment by sk4ry** [comment]
Ok, effectively it makes sense :). Thank you for the tip.

**Pull request closed without merging**